torchvision/csrc/cpu/image/writejpeg_cpu.cpp

```diff
@@ -0,0 +1,143 @@
+#include "writejpeg_cpu.h"
+
+#include <setjmp.h>
+#include <string>
+
+#if !JPEG_FOUND
+
+torch::Tensor encodeJPEG(const torch::Tensor& data, int64_t quality) {
+  TORCH_CHECK(
+      false, "encodeJPEG: torchvision not compiled with libjpeg support");
+}
+
+void writeJPEG(
+    const torch::Tensor& data,
+    std::string filename,
+    int64_t quality) {
+  TORCH_CHECK(
+      false, "writeJPEG: torchvision not compiled with libjpeg support");
+}
+
+#else
+
+#include <jpeglib.h>
+#include "jpegcommon.h"
+
+struct torch_jpeg_dst_mgr {
+  struct jpeg_destination_mgr pub;
+  uint8_t* out;
+  JOCTET* buffer;
+};
+
+torch::Tensor encodeJPEG(const torch::Tensor& data, int64_t quality) {
+  // Define compression structures and error handling
+  struct jpeg_compress_struct cinfo;
+  struct torch_jpeg_error_mgr jerr;
+
+  // Define buffer to write JPEG information to and its size
+  unsigned long jpegSize = 0;
+  uint8_t* jpegBuf = NULL;
+
+  cinfo.err = jpeg_std_error(&jerr.pub);
+  jerr.pub.error_exit = torch_jpeg_error_exit;
+
+  /* Establish the setjmp return context for my_error_exit to use. */
+  if (setjmp(jerr.setjmp_buffer)) {
+    /* If we get here, the JPEG code has signaled an error.
+     * We need to clean up the JPEG object and the buffer.
+     */
+    jpeg_destroy_compress(&cinfo);
+    if (jpegBuf != NULL) {
+      free(jpegBuf);
+    }
+
+    TORCH_CHECK(false, (const char*)jerr.jpegLastErrorMsg);
+  }
+
+  // Get image info
+  int channels, height, width;
+
+  // Move tensor to CPU and cast it into uint8
+  torch::Tensor input = data.to(torch::kCPU).to(torch::kU8);
+
+  TORCH_CHECK(
+      data.dim() >= 2 && data.dim() <= 3,
+      "Input data should be a 3-dimensional or a 2-dimensional tensor");
+
+  if (data.dim() == 3) {
+    channels = data.size(0);
+    height = data.size(1);
+    width = data.size(2);
+    input = input.permute({1, 2, 0});
```

Comment:
After a second thought, I realized that this only works because in your tests your directly read from `read_jpeg`, which returns non-contiguous tensors, and when permuting it it becomes contiguous. So we do need a `.contiguous()` here to make the code more robust. Can you also add a test for a contiguous tensor?